src/main.rs

```diff
@@ -0,0 +1,157 @@
+
+    // Set up the watchdog driver - needed by the clock setup code
+    let mut watchdog = hal::Watchdog::new(pac.WATCHDOG);
+
+    // Configure the clocks
+    //
+    // The default is to generate a 125 MHz system clock
+    let clocks = hal::clocks::init_clocks_and_plls(
+        rp_pico::XOSC_CRYSTAL_FREQ,
+        pac.XOSC,
+        pac.CLOCKS,
+        pac.PLL_SYS,
+        pac.PLL_USB,
+        &mut pac.RESETS,
+        &mut watchdog,
+    )
+    .ok()
+    .unwrap();
+
+    // The single-cycle I/O block controls our GPIO pins
+    let sio = hal::Sio::new(pac.SIO);
+
+    // Set up the USB driver
+    let usb_bus = UsbBusAllocator::new(hal::usb::UsbBus::new(
+        pac.USBCTRL_REGS,
+        pac.USBCTRL_DPRAM,
+        clocks.usb_clock,
+        true,
+        &mut pac.RESETS,
+    ));
+
+    // Set up the USB Communications Class Device driver
+    let mut serial = SerialPort::new(&usb_bus);
+
+    // Set up the USB PicoTool Class Device driver
+    let mut picotool: PicoToolReset<_> = PicoToolReset::new(&usb_bus);
+
+    // Create a USB device RPI Vendor ID and on of these Product ID:
+    // https://github.com/raspberrypi/picotool/blob/master/picoboot_connection/picoboot_connection.c#L23-L27
+    let mut usb_dev = UsbDeviceBuilder::new(&usb_bus, UsbVidPid(0x2e8a, 0x000a))
+        .manufacturer("krzmaz")
+        .product("Neopixel Driver")
+        .serial_number("PICO")
+        .device_class(0) // from: https://www.usb.org/defined-class-codes
+        .build();
+
+    let pins = rp_pico::Pins::new(
+        pac.IO_BANK0,
+        pac.PADS_BANK0,
+        sio.gpio_bank0,
+        &mut pac.RESETS,
+    );
+
+    // Create a count down timer for the Ws2812 instance:
+    let timer = Timer::new(pac.TIMER, &mut pac.RESETS);
+
+    // Split the PIO state machine 0 into individual objects, so that
+    // Ws2812 can use it:
+    let (mut pio, sm0, _, _, _) = pac.PIO0.split(&mut pac.RESETS);
+
+    // Instantiate a Ws2812 LED strip:
+    let mut ws = Ws2812::new(
+        // Use pin 6 on the Raspberry Pi Pico (which is GPIO4 of the rp2040 chip)
+        // for the LED data output:
+        pins.gpio2.into_mode(),
+        &mut pio,
+        sm0,
+        clocks.peripheral_clock.freq(),
+        timer.count_down(),
+    );
+    // 128 is the default for usb implementation
+    let mut read_buf = [0u8; 128];
+    let mut buf = Vec::<u8, 5000>::new();
+
+    loop {
+        // Check for new data
+        if usb_dev.poll(&mut [&mut serial, &mut picotool]) {
+            match serial.read(&mut read_buf) {
+                Ok(count) if count > 0 => {
+                    buf.extend(read_buf.iter().take(count).copied());
+                    if buf.len() > 1 {
+                        let length: u16 = u16::from_le_bytes([buf[0], buf[1]]);
+                        // there might be the start of the next "frame" in the input buffer
+                        let overflow = buf.len() - (length as usize + 2);
```

Comment:
What if length is greater than size of buf? This will panic. Just move it into the if statement below.